pyccel/parser/syntactic.py

```diff
@@ -729,13 +730,18 @@ def fill_types(ls):
 
                 txt  = '#$ header template ' + str(tp_name)
                 txt += '(' + '|'.join(types) + ')'
-                if tp_name in templates:
+                if tp_name in tmp_templates:
                     msg = 'The template "{}" is duplicated'.format(tp_name)
                     errors.report(msg,
                                 bounding_box = (stmt.lineno, stmt.col_offset),
                                 severity='warning')
+                tmp_templates[tp_name] = hdr_parse(stmts=txt)
+                template['templates_dict'] = tmp_templates
```

Comment:
Does `tmp_templates` need saving to `template['templates_dict']` multiple times? Both `template['templates_dict']` and `tmp_templates` should reference the same object in memory once this assignment has been done once so `template['templates_dict']` should be updated automatically.

Why do we need both objects?